extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,571 @@
+   * <p/>
+   * <strong>Note: this is a single thread service.</strong>
+   */
+  final class DruidPullMessageService extends ServiceThread
+  {
+
+    private volatile List<DruidPullRequest> requestsWrite = new ArrayList<>();
+    private volatile List<DruidPullRequest> requestsRead = new ArrayList<>();
+
+    private final DefaultMQPullConsumer defaultMQPullConsumer;
+
+    public DruidPullMessageService(final DefaultMQPullConsumer defaultMQPullConsumer)
+    {
+      this.defaultMQPullConsumer = defaultMQPullConsumer;
+    }
+
+    public void putRequest(final DruidPullRequest request)
+    {
+      synchronized (this) {
+        this.requestsWrite.add(request);
+        if (!hasNotified) {
+          hasNotified = true;
+          notify();
+        }
+      }
+    }
+
+    private void swapRequests()
+    {
+      List<DruidPullRequest> tmp = requestsWrite;
+      requestsWrite = requestsRead;
+      requestsRead = tmp;
+    }
+
+    @Override
+    public String getServiceName()
+    {
+      return getClass().getSimpleName();
+    }
+
+    /**
+     * Core message pulling logic code goes here.
+     */
+    private void doPull()
+    {
+      for (DruidPullRequest pullRequest : requestsRead) {
+        PullResult pullResult;
+        try {
+          if (!pullRequest.isLongPull()) {
+            pullResult = defaultMQPullConsumer.pull(
+                pullRequest.getMessageQueue(),
+                pullRequest.getTag(),
+                pullRequest.getNextBeginOffset(),
+                pullRequest.getPullBatchSize()
+            );
+          } else {
+            pullResult = defaultMQPullConsumer.pullBlockIfNotFound(
+                pullRequest.getMessageQueue(),
+                pullRequest.getTag(),
+                pullRequest.getNextBeginOffset(),
+                pullRequest.getPullBatchSize()
+            );
+          }
+
+          switch (pullResult.getPullStatus()) {
+            case FOUND:
+              // Handle pull result.
+              if (!messageQueueTreeSetMap.keySet().contains(pullRequest.getMessageQueue())) {
+                messageQueueTreeSetMap.putIfAbsent(
+                    pullRequest.getMessageQueue(),
+                    new ConcurrentSkipListSet<>(new MessageComparator())
+                );
+              }
+              messageQueueTreeSetMap.get(pullRequest.getMessageQueue()).addAll(pullResult.getMsgFoundList());
+              break;
+
+            case NO_NEW_MSG:
+            case NO_MATCHED_MSG:
+              break;
+
+            case OFFSET_ILLEGAL:
+              LOGGER.error(
+                  "Bad Pull Request: Offset is illegal. Offset used: {}",
+                  pullRequest.getNextBeginOffset()
+              );
+              break;
+
+            default:
+              break;
+          }
+        }
+        catch (MQClientException | RemotingException | MQBrokerException | InterruptedException e) {
+          LOGGER.error("Failed to pull message from broker.", e);
```

Comment:
You mean first wrap a new Exception instance then log down the new exception paradigm? I guess this is a kind of coding style issue.